src/ArcherContainer.js

```diff
@@ -271,20 +307,89 @@ export class ArcherContainer extends React.Component<Props, State> {
             endingPoint={endingPoint}
             endingAnchorOrientation={endingAnchorOrientation}
             strokeColor={strokeColor}
-            arrowLength={arrowLength}
             strokeWidth={strokeWidth}
             strokeDasharray={strokeDasharray}
             arrowLabel={label}
-            arrowThickness={arrowThickness}
             arrowMarkerId={this._getMarkerId(source, target)}
             noCurves={!!noCurves}
             offset={offset}
+            endShape={endShape}
           />
         );
       },
     );
   };
 
+  _buildShape = (
+    style: LineType,
+  ): {
+    markerHeight: number,
+    markerWidth: number,
+    path: React$Node,
+    refX: number,
+    refY: number,
+  } => {
+    const chosenEndShape = getEndShapeFromStyle(style);
+    const getProp = (shape: ValidShapeTypes, prop: string) => {
+      return (
+        style.endShape?.[shape]?.[prop] ||
+        this.props.endShape?.[shape]?.[prop] ||
+        ArcherContainer.defaultProps.endShape[shape][prop]
+      );
+    };
+
+    const shapeMap = {
+      circle: () => {
+        const radius = getProp('circle', 'radius');
+        const strokeWidth = getProp('circle', 'strokeWidth');
+        const strokeColor = getProp('circle', 'strokeColor');
+        const fillColor = getProp('circle', 'fillColor');
+
+        return {
+          markerWidth: radius * 4,
+          markerHeight: radius * 4,
+          refX: radius * 2 + strokeWidth,
+          refY: radius * 2,
+          path: (
+            <circle
+              cx={radius * 2}
+              cy={radius * 2}
+              r={radius}
+              fill={fillColor}
+              stroke={strokeColor}
+              strokeWidth={strokeWidth}
+            />
+          ),
+        };
+      },
+      arrow: () => {
+        const strokeColor = style.strokeColor || this.props.strokeColor;
+
+        const arrowLength =
+          style.endShape?.arrow?.arrowLength ??
+          this.props.endShape?.arrow?.arrowLength ??
+          ArcherContainer.defaultProps.endShape.arrow.arrowLength;
```

Comment:
This is the same check as we were doing before, with reference to this comment:

`// Actual arrowLength value might be 0, which can't work with a simple 'actualValue || defaultValue'`

By using the `nullish coalescing operator` (`??`) we are able to check if the left side of the operator is "nullish" (null/undefined). This is subtly different than the `logical or operator` (`||`). More details can be [found here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator#Relationship_with_the_optional_chaining_operator_.).